app/assets/javascripts/filtering_js.js

```diff
@@ -0,0 +1,10 @@
+$(function(){
+  $("#search_box").keyup(function(){
+    if(!$(this).val()) {
+      $("#top-wrapper div.lesson-card").show();
+    } else {
+      $("#top-wrapper div.lesson-card").hide();
+      $("#top-wrapper div.lesson-card h3:contains(" + this.value + ")").parent().parent().parent().show();
```

Comment:
`テンプレートリテラル` を使用しましょう。

```suggestion
      $(`#top-wrapper div.lesson-card h3:contains(${this.value})`).parent().parent().parent().show();
```

なお，`app/views/texts/index.html.erb` をh5タグにするなら，こちらも変更が必要です。

さらに，大文字・小文字の違いを無視できるようにできればなおよいですね。

Reply:
ユーザーからの入力は大文字、小文字を問わず、.toLowerCase()で強制的に小文字の検索ワードにしてしまうことで目処はついたと考え、これで実装してみました。

たけさんのサジェストが、h3:containsが`Rails`だろうと、ユーザーが`rails, RAILS, railS, raiLs, raiLs, raiLs, raIls, rAils, ... (略) ..., RAILs ` でも引っかかるようにすると勘ぐったならば、この場合はどのようになるのでしょうか？
考えた案としては、
- DOM の <%= text.title %>をすべて小文字にしたDOM を一時的に作って、上記の.toLowerCase と合致させる。
- ユーザーからの入力が`rails`だろうと、5 文字 x 2 (Upper or Lower) = 10 ケースの検索ワードのarray を作って、このarrayが検索ワードとして総当たりで検索する。
- java にあるというequalsIgnoreCase に相当する関数を自作する。

Reply:
5 文字 x 2 (Upper or Lower) = 10 ケース じゃないですね。。。
5P2 = 5 x 4 = 20 通りのワードを含むarrayですね。。。

Reply:
片方だけではなく，両方小文字に変換した前提で比較するのがポイントです。

Reply:
以下のfunction を組み合わせて実装しようとしています。
- node.textContent の 小文字化したものを、検索窓に入力されたstring の小文字化したもので index0f 評価した結果が -1 よりも大きければ、含まれる。
- group.filter 関数で ブロック内がtrue と評価されたものを抽出し、これを新しいgroup として上書きする。
- node.toggle(true) で display: block; と display: none; を切替する。

以下のコードで希望するfiltering ができると思いましたが、filter で抽出された groupの各segment を起点に parentNodeメソッドで該当する親要素node を取得できない、ことがわかりました。

たけさんのヒントをいただきたいのですが
filter function を生かしながら、親要素を取得する方法があれば教えていただきたいです。
よろしくお願いいたします。
```
$(function(){
  $("#search_box").keyup(function(){
    var myInput = $(this).val().toLowerCase();
    if(!myInput) {
      $("#top-wrapper div.lesson-card").show();
    } else {
      $("#top-wrapper div.lesson-card").hide();
      
      $("#top-wrapper div.lesson-card h3").filter(function (){
        var elem = $("#top-wrapper div.lesson-card h3");
        $(this).parentNode.parentNode.parentNode.toggle(elem.textContent.toLowerCase().indexOf(myInput) > -1);
      });
    }
  });
});
```

Reply:
`var elem = $(this).text()` ですね。あとは `parents` を使えるといい感じになると思います。

Reply:
```
$(document).on('turbolinks:load', function() { 
  $("#search_box").keyup(function(){
    var myInput = $(this).val().toLowerCase();
    if(!myInput) {
      $("#top-wrapper div.lesson-card").show();
    } else {
      $("#top-wrapper div.lesson-card").hide();
      
      $("#top-wrapper div.lesson-card h3").filter(function (){
        var elem = $(this).text().toLowerCase();
        $(this).parents('div.lesson-card').toggle(elem.indexOf(myInput) > -1);
      });
    }
  });
});
```

Reply:
filter 関数での実装例のご提示ありがとうございます。
`parents('div.lesson-card')`が一度も発想に至らなかったのは、DOM操作にまだまだ不慣れなことの証左ですね。

DOM の取り扱い方がなんとなくの理解でコードを書いていましたが、each メソッドで書いたあとに、DOMについて少し理解が進んだ気がします。